### ISSUE TITLE
dproxy not work on instruments

### ISSUE THREAD

**Issue by GGoWhere** [danielpaulus/go-ios#29]
I want to use "ios dproxy" to dump communication  between  instruments and my iphone12, but no message return. Does it work?

**Comment by nanoscopic** [comment]
It is my current understanding that the debug proxy only works with simulators, as there is less encryption going on there.

**Comment by GGoWhere** [comment]
> It is my current understanding that the debug proxy only works with simulators, as there is less encryption going on there.

But when using simulators, it cannot recognize the device. it returns:
./ios  dproxy
{"level":"fatal","msg":"no iOS devices are attached to this host","time":"2021-04-07T11:20:46+08:00"}

**Comment by danielpaulus** [comment]
No actually dproxy works only for real devices and it should be capable of dumping all communication. 
Right now the tool is probably a little hard to use. It creates a bunch of json files in directories for every connection the mac made.:

`ls -R dump*`
returns:
```
connection-#1-2021.03.30-06.06.31.045 connection-#2-2021.03.30-06.06.31.046 connections.json

dump-2021.03.30-06.06.18.914/connection-#1-2021.03.30-06.06.31.045:
jsondump-fromdevice.bin jsondump-todevice.bin

dump-2021.03.30-06.06.18.914/connection-#2-2021.03.30-06.06.31.046:
jsondump-fromdevice.bin jsondump-todevice.bin
```
Let me know if that works for you. I am also happy to make it easier to use. The complicated file structure is due to the original plan of building a graphical tool around it.

**Comment by GGoWhere** [comment]
> No actually dproxy works only for real devices and it should be capable of dumping all communication.
> I have also this one: https://github.com/danielpaulus/ios_simulator_dtx_dump for dumping simulator XCUITest communication. Simulators expose a unix domain socket for testmanagerd that also uses DTX. That was a bit easier in the beginning to get stuff working and build the DTX codec :-)
> 
> Right now the dproxy tool is probably a little hard to use. It creates a bunch of json files in directories for every connection the mac made.:
> For example here is one of my runs:
> `ls -R dump*`
> returns:
> 
> ```
> connection-#1-2021.03.30-06.06.31.045 connection-#2-2021.03.30-06.06.31.046 connections.json
> 
> dump-2021.03.30-06.06.18.914/connection-#1-2021.03.30-06.06.31.045:
> jsondump-fromdevice.bin jsondump-todevice.bin
> 
> dump-2021.03.30-06.06.18.914/connection-#2-2021.03.30-06.06.31.046:
> jsondump-fromdevice.bin jsondump-todevice.bin
> ```
> 
> Let me know if that works for you. I am also happy to make it easier to use. The complicated file structure is due to the original plan of building a graphical tool around it.
> 
> Also let me know if you are missing a feature in go-ios or open an issue for the feature that is missing :-)
> If you are looking for CPU, memory etc. I have built that before and have a pretty good idea how to do it.

You're doing a great job. Yeah, I've finished getting the CPU, memory by using IDA with dtx plugin on my iphone8(ios 12).But now I want to get more info like gpu bandwidth on my iphone12(ios14), the IDA plugin doesn't work.
So I found your tool. There's also a tool which is named perfdog like instruments. Go-ios can dump the communication text when using perfdog. It's amazing, your tool can work on the phone that not jailbreak :-). However, your go-ios can not work on the instruments. If I need gpu bandwidth, I have to use instruments instead of perfdog. 
I want to know if it's my computer environment that makes the tool unusable?

**Comment by nanoscopic** [comment]
I implemented instruments for fetching memory and cpu usage in iosif. The same technique used there can be ported ( rewritten in golang based on what it does ) to go-ios.

There is no need to mess with IDA to do it.

There is no need for paid tools like perfdog. The technique to do so is known.

**Comment by GGoWhere** [comment]
> I implemented instruments for fetching memory and cpu usage in iosif. The same technique used there can be ported ( rewritten in golang based on what it does ) to go-ios.
> 
> There is no need to mess with IDA to do it.
> 
> There is no need for paid tools like perfdog. The technique to do so is known.
> 
> See https://github.com/nanoscopic/iosif/blob/main/service_sysmon.c
> 
> Note also https://github.com/nanoscopic/iosif/blob/main/archiver/unarchiver.c#L116
> Messages from sysmontap are compressed. Lz4 is needed to decompress them.

Great!
I've also got the cpu/memory/sysmontap information.It's not easy.
Now, I'm looking for a way to get the "gpu bandwidth" information by instruments.It seems that IDA and go-ios not work.

**Comment by nanoscopic** [comment]
I saw the gpu calls when I was digging through the available functions... I didn't bother to write the functionality yet though. There are also no examples of how it works so you'll need to do as you suggest and dump the log of something else that does it.

When you figure out how the gpu stuff works let us know and we can add it.

**Comment by danielpaulus** [comment]
Ok  @GGoWhere  let me check out tonight why it is breaking and send a fix

**Comment by GGoWhere** [comment]
> Ok @GGoWhere let me check out tonight why it is breaking and send a fix

Thanks!

Waiting for good news :-)

**Comment by GGoWhere** [comment]
> I saw the gpu calls when I was digging through the available functions... I didn't bother to write the functionality yet though. There are also no examples of how it works so you'll need to do as you suggest and dump the log of something else that does it.
> 
> When you figure out how the gpu stuff works let us know and we can add it.

Yeah，I hope that go-ios can dump the text when instruments is getting "gpu bandwidth" info, so I can find out how it work.
But go-ios can not dump any texts when instruments works on my computer.
I need your help!

**Comment by danielpaulus** [comment]
I was able to reproduce issues. Working on a fix.

**Comment by danielpaulus** [comment]
It seems like something they introduced in iOS 14, it still works with older devices.

**Comment by GGoWhere** [comment]
> It seems like something they introduced in iOS 14, it still works with older devices.
> I was able to get it working again. Will hopefully soon merge the PR.

Wow, That's good news.Thank you so much!
 
How long will it take to wait for the release version? I can't wait to use it.
Thanks again!

**Comment by GGoWhere** [comment]
> It seems like something they introduced in iOS 14, it still works with older devices.
> I was able to get it working again. Will hopefully soon merge the PR.

Hi, I successfully builded the branch you fix #30.
And it really print something when Instruments working.
But the text it print seems too simple.like this:
<img width="618" alt="截屏2021-04-12 下午6 52 11" src="https://user-images.githubusercontent.com/75970030/114384420-45625e00-9bc1-11eb-9ab5-bfb8320161cd.png">
If I use the previous version when perfodg working, it print like this:
<img width="566" alt="截屏2021-04-12 下午6 54 09" src="https://user-images.githubusercontent.com/75970030/114384572-75116600-9bc1-11eb-826e-d0224f0ddbe5.png">
the previous version can dump the message name and the args and so on.
Is there somthing wrong with my environment?

**Comment by GGoWhere** [comment]
> > It seems like something they introduced in iOS 14, it still works with older devices.
> > I was able to get it working again. Will hopefully soon merge the PR.
> 
> Hi, I successfully builded the branch you fix #30.
> And it really print something when Instruments working.
> But the text it print seems too simple.like this:
> <img alt="截屏2021-04-12 下午6 52 11" width="618" src="https://user-images.githubusercontent.com/75970030/114384420-45625e00-9bc1-11eb-9ab5-bfb8320161cd.png">
> If I use the previous version when perfodg working, it print like this:
> <img alt="截屏2021-04-12 下午6 54 09" width="566" src="https://user-images.githubusercontent.com/75970030/114384572-75116600-9bc1-11eb-826e-d0224f0ddbe5.png">
> the previous version can dump the message name and the args and so on.
> Is there somthing wrong with my environment?

Oh, sorry, my fault.
I checked the files I downloaded,  and I found out that these files weren't the ones you committed!#30
Maybe I didn't use github well enough. Forgive me. Orz.
I'll try to replace the files you committed and re-build them.

**Comment by GGoWhere** [comment]
Thank you very much! 
Seeing your latest submission, it seems to have some effect, but there are still some problems.
Maybe I can open a new issue
